File: leet-code/3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters3_2.java
import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeatingCharacters3_2 {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int maxSubstringLength = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            // 해시맵에 문자가 존재하고 해시맵의 값이 left 범위 안에 있는 경우
            if (map.containsKey(c) && map.get(c) >= left) {
                left = map.get(c) + 1;
            } else {
                maxSubstringLength = Math.max(maxSubstringLength, right - left + 1);
            }
            map.put(c, right);
        }
        return maxSubstringLength;
    }
}

File: leet-code/3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters3_2.kt
package com.example

class LongestSubstringWithoutRepeatingCharacters3_2 {
    fun lengthOfLongestSubstring(s: String): Int {
        val map = mutableMapOf<Char, Int>()
        var left = 0
        var maxLength = 0
        for (right in s.indices) {
            if (map[s[right]] != null && map[s[right]]!! >= left) {
                left = map[s[right]]!! + 1
            } else {
                maxLength = maxLength.coerceAtLeast(right - left + 1)
            }
            map[s[right]] = right
        }
        return maxLength
    }
}
